### PR TITLE
Arming check function

### PR DESCRIPTION
**Pull request by peterbarker**
I'm after feedback on this check() function.

This accomplishes two things
 - allows us to sneak up on the always-do-arming-checks patches from a different direction
 - will allow us to easily have a bitmask of arming checks which have failed

Once concern I have is the signature of the check function; the first argument is a boolean which says if the check is passing, the third is a message to emit if the check fails.  that feels a little inconsistent, but having check take a failure boolean doesn't seem right, and I can't see ways to reword the error messages satisfactorily.

### CODE REVIEW

File: libraries/AP_Arming/AP_Arming.cpp [hunk 1]
```diff
@@ -111,59 +111,62 @@ uint16_t AP_Arming::get_enabled_checks()
...
-            return false;
-        }
-    }
-    return true;
+    return (
+        check(!DataFlash_Class::instance()->logging_failed(),
+              report,
+              ARMING_CHECK_LOGGING,
+              "Logging failed") &
+        check(DataFlash_Class::instance()->CardInserted(),
```
Comment (khancyr) [inline libraries/AP_Arming/AP_Arming.cpp]: doesn't it miss the negation ?

**Comment by khancyr** [comment]
it looks good to me but I agree that the logic is a little weird. I don't have another idea for now

**Comment by OXINARF** [comment]
I don't like it. I understand it reduces code repetition a lot, but code looks much more confusing to me.

Maybe smaller steps would be better? `check_enable` looks fine to add (although a macro that would just get the checking bit but do the `return true` part too would reduce more code). Maybe another macro doing something like `if(report) { gcs().send_message(CRITICAL, <argument>); }`?

**Comment by peterbarker** [comment]
Amusingly, I started off by adding the `check_enabled` function - and went through and used it everywhere!  This was part of the "always run arming checks" PR, 'though - so subsequent patches removed all of those lines!

Do you have issues with `check()`, or just the way I used it?  I could make the usage more verbose,  perhaps?

**Comment by OXINARF** [comment]
> Do you have issues with check(), or just the way I used it? I could make the usage more verbose, perhaps?

I have two issues with it:

- checks are run, even if they are disabled, they just don't stop arming or generate a message
- code is harder to read, the difference between & and && is very subtle and then there are things like https://github.com/peterbarker/ardupilot/blob/b2e5cf1c2226ccc4a6c2f43e5fdc3c60c2fb2adb/libraries/AP_Arming/AP_Arming.cpp#L334-L343 that doesn't help too

**Comment by peterbarker** [comment]
On Fri, 28 Jul 2017, Francisco Ferreira wrote:

>  *  checks are run, even if they are disabled, they just don't stop arming or generate a message

That's the plan anyway with the always-run-arming-checks branch - they 
won't stop arming, but will generate a message at a lower priority :-)

>  *  code is harder to read, the difference between & and && is very subtle and then there are things like https://github.com/peterbarker/ardupilot/blob/b2e5cf1c2226ccc4a6c2f43e5fdc3c60c2fb2adb/libraries/AP_Arming/AP_Arming.cpp#L334-L343 that doesn't help too

The former ones I can fix.

The GPS one is harder.

I might put this aside for now and try to revisit it later.

**Comment by peterbarker** [comment]
Thanks for the feedback!

**Comment by OXINARF** [comment]
@peterbarker 

> That's the plan anyway with the always-run-arming-checks branch - they
won't stop arming, but will generate a message at a lower priority :-)

I can agree to that if all checks are run (which they aren't with this version) and we log what passed and failed - if we don't log, what is the point in running them all?
By the way, I'd love to end the split between pre-arming and arming checks.

> The former ones I can fix.
>
> The GPS one is harder.

Yeah, what I meant is that it doesn't help (me at least) to have a clear look at the code, it actually makes the opposite, harder to know what is happening.

> I might put this aside for now and try to revisit it later.

Thanks, I like refactoring, just not this one 🙂

**Pull request closed without merging**